extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -188,10 +187,22 @@ private Table readTableInternal(
         // Load the partitioning schema
         final org.apache.iceberg.PartitionSpec partitionSpec = table.spec();
 
-        // Use the user-supplied defininition or the Iceberg schema for the table definition.
-        final TableDefinition tableDef = instructions.tableDefinition().isPresent()
-                ? instructions.tableDefinition().get()
-                : fromSchema(schema, partitionSpec, instructions);
+        // Get the user supplied table definition and column rename map (if any).
+        final TableDefinition userTableDef = instructions.tableDefinition().orElse(null);
+
+        // Get the table definition from the schema (potentially limited by the user supplied table definition and
+        // applying column renames).
+        final TableDefinition icebergTableDef =
+                fromSchema(schema, partitionSpec, userTableDef, instructions.columnRenameMap());
+
+        // If the user supplied a table definition, make sure it's fully compatible.
+        final TableDefinition tableDef;
+        if (userTableDef != null) {
+            tableDef = icebergTableDef.checkMutualCompatibility(userTableDef);
```

Comment:
We additionally want to test that all user columns marked as partitioning are also so marked in the iceberg def.
I think it is good/ok to have *extra* partitioning columns (that is, the user wants to see them, didn't mark them as partitioning, Iceberg says they are partitioning, and our result says they are partitioning). We should check this with Andy + Charles to make sure we're all on the same page.